addons/report/models/report.py

```diff
@@ -198,72 +469,54 @@ def get_pdf(self, docids, report_name, html=None, data=None):
-        irconfig_obj = self.env['ir.config_parameter'].sudo()
-        base_url = irconfig_obj.get_param('report.url') or irconfig_obj.get_param('web.base.url')
-
-        # Minimal page renderer
-        view_obj = self.env['ir.ui.view']
-        render_minimal = partial(view_obj.with_context(context).render_template, 'report.minimal_layout')
-
-        # The received html report must be simplified. We convert it in a xml tree
-        # in order to extract headers, bodies and footers.
-        try:
-            root = lxml.html.fromstring(html)
-            match_klass = "//div[contains(concat(' ', normalize-space(@class), ' '), ' {} ')]"
-
-            for node in root.xpath(match_klass.format('header')):
-                body = lxml.html.tostring(node)
-                header = render_minimal(dict(subst=True, body=body, base_url=base_url))
-                headerhtml.append(header)
+        headerhtml, contenthtml, footerhtml, sorted_docids, specific_paperformat_args = self.with_context(context)._prepare_html(html, docids, report.model)
 
-            for node in root.xpath(match_klass.format('footer')):
-                body = lxml.html.tostring(node)
-                footer = render_minimal(dict(subst=True, body=body, base_url=base_url))
-                footerhtml.append(footer)
+        # Run wkhtmltopdf process
+        documents_paths = run_wkhtmltopdf(
+            headerhtml, footerhtml, contenthtml, context.get('landscape'),
+            paperformat, specific_paperformat_args,
+            context.get('set_viewport_size'),
+        )
 
-            for node in root.xpath(match_klass.format('article')):
-                # Previously, we marked some reports to be saved in attachment via their ids, so we
-                # must set a relation between report ids and report's content. We use the QWeb
-                # branding in order to do so: searching after a node having a data-oe-model
-                # attribute with the value of the current report model and read its oe-id attribute
-                if docids and len(docids) == 1:
-                    reportid = docids[0]
+        # Look for a post_process_report function
+        if not post_process_report:
+            post_process_report = context.get('post_process_report') or report.post_process_report
+
+        # Iterate through documents to post_process_report the generated reports
+        # in case of the record inherits report.mixin
+        report_content = None
+        if post_process_report:
+            records = None
+            for it, docid in enumerate(sorted_docids):
+                # Skip if no docid found
+                if not docid:
+                    continue
+                with open(documents_paths[it], 'rb') as pdfdocument:
+                    report_content = pdfdocument.read()
+                record = self.env[report.model].browse(docid)
+                report_content = getattr(record, post_process_report)(report_content)
```

Comment:
@smetl How can we know when the post_process method is called on the generated document for one record ( :champagne: )  or the aggregated result if the same method is called in both cases?
Can we imagine the following approach:
* A default post_process method on the report
* The ability to specify a prost_process_report method in the report definition to call on the model with the content AND the report as parameter. If the method is specified, this method is called in place of the method on the report.

To dissociate the case where one of the 2 methods (the one on the report or the specialized one on the model recordset) is called on the aggregated result or the result for one model instance  we could use an additional parameter.

regards,

lmi